drivers/kw41zrf/kw41zrf_getset.c

```diff
@@ -0,0 +1,296 @@
+ * details.
+ */
+
+/**
+ * @ingroup     drivers_kw41zrf
+ * @{
+ * @file
+ * @brief       get/set functionality of kw41zrf driver
+ *
+ * @author  Joakim Nohlgård <joakim.nohlgard@eistec.se>
+ * @}
+ */
+
+#include <errno.h>
+#include "log.h"
+#include "cpu.h"
+#include "byteorder.h"
+#include "kw41zrf.h"
+#include "kw41zrf_intern.h"
+#include "kw41zrf_getset.h"
+
+#define ENABLE_DEBUG (0)
+#include "debug.h"
+
+#define KW41ZRF_NUM_CHANNEL      (KW41ZRF_MAX_CHANNEL - KW41ZRF_MIN_CHANNEL + 1)
+
+/* Lookup table for PA_PWR register */
+/* Source: KW41Z data sheet, 5.3 Transmit and PLL Feature Summary,
+ * Table 8. Transmit Output Power as a function of PA_POWER[5:0] */
+static const uint8_t tx_power_dbm_to_pa_pwr[29] = {
+    4, 4, 4, 4,        /* -19:-16 dBm */
+    6, 6,              /* -15:-14 dBm */
+    8, 8,              /* -13:-12 dBm */
+    10, 10,            /* -11:-10 dBm */
+    12,                /* -9 dBm */
+    14,                /* -8 dBm */
+    16,                /* -7 dBm */
+    18,                /* -6 dBm */
+    20,                /* -5 dBm */
+    22,                /* -4 dBm */
+    26,                /* -3 dBm */
+    28,                /* -2 dBm */
+    34,                /* -1 dBm */
+    38,                /* 0 dBm */
+    42,                /* 1 dBm */
+    48,                /* 2 dBm */
+    56,                /* 3 dBm */
+    62,                /* 4 dBm */
+};
+
+void kw41zrf_set_tx_power(kw41zrf_t *dev, int16_t txpower_dbm)
+{
+    if (txpower_dbm < KW41ZRF_OUTPUT_POWER_MIN) {
+        ZLL->PA_PWR = 1;
+    }
+    else if (txpower_dbm > KW41ZRF_OUTPUT_POWER_MAX) {
+        ZLL->PA_PWR = 62;
+    }
+    else {
+        ZLL->PA_PWR = tx_power_dbm_to_pa_pwr[txpower_dbm - KW41ZRF_OUTPUT_POWER_MIN];
+    }
+
+    LOG_DEBUG("[kw41zrf] set txpower to: %d\n", txpower_dbm);
+    dev->tx_power = txpower_dbm;
+}
+
+uint16_t kw41zrf_get_txpower(kw41zrf_t *dev)
+{
+    return dev->tx_power;
+}
+
+uint8_t kw41zrf_get_channel(kw41zrf_t *dev)
+{
+    (void) dev;
+    return (ZLL->CHANNEL_NUM0 & ZLL_CHANNEL_NUM0_CHANNEL_NUM0_MASK);
+}
+
+int kw41zrf_set_channel(kw41zrf_t *dev, uint8_t channel)
+{
+    if (channel < KW41ZRF_MIN_CHANNEL || channel > KW41ZRF_MAX_CHANNEL) {
+        LOG_ERROR("[kw41zrf] Invalid channel %u\n", channel);
+        return -EINVAL;
+    }
+
+    ZLL->CHANNEL_NUM0 = channel;
+    dev->netdev.chan = channel;
+
+    LOG_DEBUG("[kw41zrf] set channel to %u\n", channel);
+    return 0;
+}
+
+void kw41zrf_set_pan(kw41zrf_t *dev, uint16_t pan)
+{
+    dev->netdev.pan = pan;
+
+    ZLL->MACSHORTADDRS0 = (ZLL->MACSHORTADDRS0 & ~ZLL_MACSHORTADDRS0_MACPANID0_MASK) |
+        ZLL_MACSHORTADDRS0_MACPANID0(pan);
+
+    LOG_DEBUG("[kw41zrf] set pan to: 0x%x\n", pan);
+}
+
+void kw41zrf_set_addr_short(kw41zrf_t *dev, uint16_t addr)
+{
+#ifdef MODULE_SIXLOWPAN
+    /* https://tools.ietf.org/html/rfc4944#section-12 requires the first bit to
+     * 0 for unicast addresses */
+    addr &= 0x7fff;
```

Comment:
@bergzand @miri64 Is this necessary or should this be handled by upper layers?

Reply:
> Is this necessary

I'm afraid so

> or should this be handled by upper layers?

I would prefer this, but I haven't really checked whether handling it in a upper layer makes sense.

I'd say keep this in, I have no intention at the moment to refactor it to something else.